casper/src/main/resources/PoS.rhox

```diff
@@ -241,245 +251,269 @@ new  PoS,
+                            validatorsToWithdrawListCh,
+                            isQuarantineFinished,
+                            notWithdrawn,
+                            newBondsListCh,
+                            newWithdrawersListCh,
+                            newRewardsListCh,
+                            paymentsCh, computePay
+                        in {
+                          @ListOps!("filter", withdrawers.toList(), *isQuarantineFinished, *quarantinedValidatorsCh) |
+                          for (@quarantinedValidators <- quarantinedValidatorsCh) {
+                            @ListOps!("map", quarantinedValidators, *fst, *validatorsToWithdrawListCh) |
+                            for (@validatorsToWithdrawList <- validatorsToWithdrawListCh) {
+                              @ListOps!("map", validatorsToWithdrawList, *computePay, *paymentsCh) |
+                              contract computePay(@pk, resultCh) = {
+                                resultCh!((pk, allBonds.get(pk) + rewards.getOrElse(pk, 0)))
+                              } |
+
+                              match validatorsToWithdrawList.toSet() {
+                                validatorsToWithdrawSet => {
+                                  @ListOps!("filter", allBonds.toList(), *notWithdrawn, *newBondsListCh) |
+                                  @ListOps!("filter", withdrawers.toList(), *notWithdrawn, *newWithdrawersListCh) |
+                                  @ListOps!("filter", rewards.toList(), *notWithdrawn, *newRewardsListCh) |
+
+                                  for (@newWithdrawersList <- newWithdrawersListCh;
+                                       @newBondsList <- newBondsListCh;
+                                       @newRewardsList <- newRewardsListCh;
+                                       @payments <- paymentsCh) {
+                                    returnCh!((payments, newBondsList.toMap(), newWithdrawersList.toMap(), newRewardsList.toMap()))
+                                  } |
+
+                                  contract notWithdrawn(@(pk, _), resultCh) = { resultCh!(not validatorsToWithdrawSet.contains(pk))}
+                                }
+                              }
+                            }
+                          } |
 
-                            contract notWithdrawn(@(pk, _), resultCh) = { resultCh!(not validatorsToWithdrawSet.contains(pk))}
+                          contract isQuarantineFinished(@(_, blockNumber), resultCh) = {
+                            resultCh!(blockNumber >= currentBlockNumber)
                           }
                         }
                       }
-                    } |
-
-                    contract isQuarantineFinished(@(_, blockNumber), resultCh) = {
-                      resultCh!(blockNumber >= currentBlockNumber)
                     }
+                  } else {
+                    ackCh!((false, "Invalid system auth token"))
```

Comment:
`closeBlock` is returning `Nil` for success. Maybe it makes sense to change `slash` and `closeBlock` to return Either as other methods.